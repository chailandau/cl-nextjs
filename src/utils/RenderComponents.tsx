import { FC, Fragment } from 'react';

import { Page_PageSections, Project_ProjectSections } from '@/api/graphqlTypes';
import Image from '@/atoms/Image';
import Hero from '@/components/Hero';
import SingleUse from '@/components/SingleUse';
import Socials from '@/components/Socials';
import TextBlock from '@/molecules/TextBlock/TextBlock';

interface Sections {
    /** Sections to render */
    components?: Page_PageSections[] | Project_ProjectSections[];
}

const RenderComponents: FC<Sections> = ({ components }) => {
    if (!components || !components.length) {
        return null;
    }
    const renderComponent = (
        component: Page_PageSections | Project_ProjectSections
    ) => {
        switch (component?.__typename) {
            case 'HeroBlock':
                return <Hero {...component?.hero} />;
            case 'TextBlock':
                return (
                    <TextBlock
                        heading={component?.heading}
                        richText={component?.text}
                    />
                );
            case 'ImageBlock':
                return (
                    <Image
                        src={component?.image?.url || ''}
                        alt={component?.image?.alt || ''}
                        width={component?.image?.width || 1000}
                        height={component?.image?.height || 1000}
                    />
                );
            case 'SocialsBlock':
                return <Socials socials={component?.socials} />;
            case 'SingleUseBlock':
                return <SingleUse component={component?.singleUse} />;
            default:
                return null;
        }
    };

    return (
        <>
            {components?.map((component) => (
                <Fragment key={component?.id}>
                    {renderComponent(component)}
                </Fragment>
            ))}
        </>
    );
};

export default RenderComponents;
