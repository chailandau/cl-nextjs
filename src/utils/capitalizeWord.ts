/**
 * Capitalizes the first letter of a string.
 *
 * @param {string} str - String to be capitalized.
 * @return {string} Capitalized string.
 */
export const capitalizeWord = (str: string): string => {
    if (!str) {
        return str;
    }

    const firstLetter = str.charAt(0).toUpperCase();
    const restOfWord = str.slice(1);

    return firstLetter + restOfWord;
};
