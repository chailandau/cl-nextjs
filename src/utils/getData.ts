import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

export const getData = async (QUERY: string) =>
    await graphQLClient.request<Query>(QUERY);

export const getPageData = async (QUERY: string, slug: string) =>
    await graphQLClient.request<Query>(QUERY, { slug });
