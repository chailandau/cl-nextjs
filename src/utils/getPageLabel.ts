import { Nav_MenuItems } from '@/api/graphqlTypes';

export const getPageLabel = (menuItem: Nav_MenuItems) => {
    const customLabel =
        menuItem?.overridePageName &&
        menuItem?.customLabel !== null &&
        menuItem.customLabel;

    return menuItem?.page && (customLabel || menuItem?.page?.title);
};
