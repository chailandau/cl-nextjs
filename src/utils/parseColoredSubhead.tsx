import { JSX } from 'react';

import Text, { TextProps } from '@/atoms/Text';

interface ChildSegment {
    text: string;
    bold?: boolean;
}

export interface ColoredSubheadProps {
    children: ChildSegment[];
}

type ColoredSubhead = JSX.Element[] | null | undefined;

/**
 * Parses a colored subhead and returns an array of Text components with appropriate styling.
 *
 * @param coloredSubhead - An array of objects representing the colored subhead.
 * @returns An array of Text components with appropriate styling, or null if the input is falsy.
 */
export const parseColoredSubhead = (
    coloredSubhead: ColoredSubheadProps[]
): ColoredSubhead | null => {
    if (!coloredSubhead) {
        return null;
    }

    const colors = ['aqua', 'white', 'pink'];

    return coloredSubhead.flatMap((segment) => {
        const boldIndexes: number[] = [];

        return segment?.children?.map((child, index) => {
            const isBold = child?.bold;
            let textBgColor;

            if (isBold) {
                boldIndexes.push(index);
                const position = boldIndexes?.indexOf(index);
                textBgColor = colors[position % colors?.length];
            }

            const textProps = {
                as: 'span' as TextProps['as'],
                size: 'md' as TextProps['size'],
                font: 'secondary' as TextProps['font'],
                backgroundColor: textBgColor
            };

            return (
                <Text key={index} {...textProps}>
                    {child?.text}
                </Text>
            );
        });
    });
};
