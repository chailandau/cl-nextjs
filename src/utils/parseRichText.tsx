import { JSX } from 'react';

import styles from '../atoms/Text/Text.module.scss';

import Text from '@/atoms/Text';
import { textSizes } from '@/atoms/Text/Text';

interface ChildSegment {
    text: string;
}

export interface RichTextProps {
    children: ChildSegment[];
}

type RichText = JSX.Element[] | null | undefined;

/**
 * Parses rich text and returns an array of Text components.
 *
 * @param richText - An array of objects representing rich text.
 * @returns An array of Text components, or null if the input is falsy.
 */
export const parseRichText = (
    richText: RichTextProps[],
    size: (typeof textSizes)[number] = 'xs'
): RichText | null => {
    if (!richText) {
        return null;
    }

    const results: JSX.Element[] = [];

    richText.flatMap(
        (segment) =>
            segment?.children?.forEach((child) => {
                const paragraphs = child.text
                    .split('\n')
                    .filter((paragraph) => paragraph.trim() !== '');

                paragraphs.forEach((par, parIndex) => {
                    results.push(
                        <Text
                            className={styles['rich-text']}
                            size={size}
                            key={parIndex}
                        >
                            {par}
                        </Text>
                    );
                });
            })
    );

    return results;
};
