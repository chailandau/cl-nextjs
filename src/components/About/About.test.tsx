import { composeStory } from '@storybook/react';
import { render, screen } from '@testing-library/react';

import Meta, { Default } from './About.stories';

import { testAxeAndSnapshot } from '@/utils/testHelpers';

const About = composeStory(Default, Meta);

describe('About', () => {
    it('defaults correctly', () => {
        render(<About data={undefined} />);
        expect(screen.queryByRole('section')).not.toBeInTheDocument();
    });
    testAxeAndSnapshot({
        component: <About />
    });
});
