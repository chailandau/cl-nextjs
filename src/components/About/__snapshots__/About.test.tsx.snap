// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="flex section about"
  >
    <div
      class="image"
    >
      <div
        class="flex container border"
      >
        <img
          alt="Chai Landau smiling, surrounded by greenery"
          data-nimg="1"
          decoding="async"
          height="1191"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fabout-pic.jpg&w=2048&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fabout-pic.jpg&w=1080&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fabout-pic.jpg&w=2048&q=75 2x"
          style="color: transparent;"
          width="1000"
        />
      </div>
      <div
        class="pets"
      >
        <div
          class="flex container warren"
        >
          <img
            alt="Warren - greyhound with tongue sticking out of side of mouth"
            data-nimg="1"
            decoding="async"
            height="181"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fwarren.png&w=640&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fwarren.png&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fwarren.png&w=640&q=75 2x"
            style="color: transparent;"
            width="270"
          />
        </div>
        <div
          class="flex container harvey"
        >
          <img
            alt="Harvey - orange cat with yellow eyes"
            data-nimg="1"
            decoding="async"
            height="149"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fharvey.png&w=384&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fharvey.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fharvey.png&w=384&q=75 2x"
            style="color: transparent;"
            width="144"
          />
        </div>
        <div
          class="flex container athena"
        >
          <img
            alt="Athena - brown tabby cat"
            data-nimg="1"
            decoding="async"
            height="120"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fathena.png&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fathena.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Fimages%2Fathena.png&w=256&q=75 2x"
            style="color: transparent;"
            width="120"
          />
        </div>
      </div>
    </div>
    <div
      class="flex content"
    >
      <h1
        class="heading lg primary"
      >
        Hello! I’m Chai.
      </h1>
      <div
        class="colored-subhead subhead"
      >
        <span
          class="text md secondary"
        >
          I’m a 
        </span>
        <span
          class="text md bg-aqua secondary"
        >
          tinkerer
        </span>
        <span
          class="text md secondary"
        >
           at heart with a passion for creating 
        </span>
        <span
          class="text md bg-white secondary"
        >
          functional
        </span>
        <span
          class="text md secondary"
        >
           and 
        </span>
        <span
          class="text md bg-pink secondary"
        >
          delightful
        </span>
        <span
          class="text md secondary"
        >
           user experiences.
        </span>
      </div>
      <div>
        <p
          class="text sm primary rich-text"
        >
          I prioritize accessibility and functionality in everything I do, whether it is sewing, woodworking, or architecting a component library. 
        </p>
        <p
          class="text sm primary rich-text"
        >
          My design background ensures that my code choices are well-informed and part of a bigger picture.
        </p>
        <p
          class="text sm primary rich-text"
        >
          When I’m not customizing my mechanical keyboard (Keychron K1 Pro, brown switch, RGB backlighting), chances are I am messing around on my Linux server, daydreaming about my next sewing project, or enjoying quality time with my partner and pets.
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
